### PR TITLE
Feat: #90 #92 #94 SQL template management

### PR DESCRIPTION
**Pull request by fedorgrab**
Establishing a SQL Template generating library. This PR partially addresses issue #94. Additionally, it resolves issues #92, and #90 (which was previously closed due to the older implementation of filtering). 
However, not all SQL queries have been refactored; this task should be continued under #94.

Below is an example of the output log produced by an infrastructure test after applying these changes:

```
INFO     casp.cell_data_manager.sql.query:query.py:46 Rendered SQL Query:
create or replace table `dsp-cell-annotation-service.cas_test_dataset.test_extract_homo_sap__extract_cell_info` partition by range_bucket(extract_bin, generate_array(0, 40000, 10)) cluster by extract_bin as
select cas_cell_index,
       cas_ingest_id,
       cell_type,
       total_mrna_umis,
       donor_id,
       assay,
       development_stage,
       disease,
       organism,
       sex,
       tissue,
       dataset_filename,
       cast(floor((row_number() over () - 1) / 10000) as int) as extract_bin
from `dsp-cell-annotation-service.cas_test_dataset.test_extract_homo_sap__extract_cell_info_randomized` c
```

This demonstrates how `casp/cell_data_manager/sql.templates/prepare_curriculum/prepare_cell_info.sql.mako` was rendered during the extraction phase.

### CODE REVIEW

File: src/casp/bq_scripts/prepare_for_training.py [hunk 1]
```diff
@@ -267,7 +159,7 @@ def prepare_extract_matrix(
     For more info on BigQuery partitioning refer to
     https://cloud.google.com/bigquery/docs/partitioned-tables
     """
-    sql = f"""
+    sql_query = f"""
```
Comment (mbabadi) [inline src/casp/bq_scripts/prepare_for_training.py:162]: Why don't you make this a template as well?

File: src/casp/cell_data_manager/README.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
+# Cell Data Manager Module
```
Comment (mbabadi) [inline src/casp/cell_data_manager/README.md:1]: Cell Data Manager Module -> Cellarium Data Manager Module

Comment (mbabadi) [reply src/casp/cell_data_manager/README.md:1]: also, rename the directory structure, `cell_data_manager` -> `cellarium_data_manager`.

File: src/casp/cell_data_manager/README.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+This module is designed for managing cell data. Currently, we store cell data in BigQuery, which uses its own SQL dialect to handle data in the warehouse.
```
Comment (mbabadi) [inline src/casp/cell_data_manager/README.md:3]: cell data -> single-cell data in Cellarium DataStore.

File: src/casp/cell_data_manager/README.md [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+Template generator is used for managing SQL queries due to the following reasons:
+
+1. `cellarium-cloud` relies on a large number of queries with dynamic content (e.g., optional joins, different filters in `where` clauses).
+2. To maintain the DRY (Don't Repeat Yourself) principle in SQL queries.
+3. To separate SQL code from Python files, which makes our codebase cleaner, more modular, and thus more easily maintainable.
+
+### Template Generator
+
+[Mako template generator](https://www.makotemplates.org/) is used for this purpose. I considered both [Mako](https://www.makotemplates.org/) and [Jinja2](https://jinja.palletsprojects.com/en/2.10.x/), which were originally designed for HTML file templating. However, they seem suitable for managing SQL queries as well. You can learn more from these references:
```
Comment (mbabadi) [inline src/casp/cell_data_manager/README.md:18]: I -> We :)

File: src/casp/cell_data_manager/README.md [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+There are also wrappers around both Jinja and Mako that are specifically used for SQL management:
+
+- [Quma (Mako wrapper)](https://github.com/ebenefuenf/quma)
+- [JinjaSQL (Jinja wrapper)](https://github.com/sripathikrishnan/jinjasql)
+
+However, it appears that these libraries are not well-maintained and/or have few stars on GitHub.
+
+### Why Mako?
+
+In my personal opinion, [Mako template generator](https://www.makotemplates.org/) seems to have a better syntax. It is easier to use and has a more Pythonic approach. It also allows for the seamless integration of Python helper functions within templates.
```
Comment (mbabadi) [inline src/casp/cell_data_manager/README.md:33]: In our opinion :)

This is not a blog post Fedor :p

File: src/casp/cell_data_manager/README.md [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+## How to use `cell_data_manager` module
```
Comment (mbabadi) [inline src/casp/cell_data_manager/README.md:35]: cellarium_datastore_manager

File: src/casp/cell_data_manager/README.md [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+2. **Parameterized Queries**: If there is a need to expose SQL queries to end-users, always use [Parameterized Queries](https://cloud.google.com/bigquery/docs/parameterized-queries#python). Parameterized queries help prevent SQL injection by separating data from SQL code.
+
+3. **Consider extending `mako_helpers`**: To simplify the implementation of parameterized queries, consider expanding `mako_helpers` to work seamlessly with [Parameterized Queries](https://cloud.google.com/bigquery/docs/parameterized-queries#python).
+
+By following these practices, you can ensure the safe and secure use of this module within the project.
+
+## To Consider
+
+Currently, there's no SQL formatting in the `casp.cell_data_manager.sql` module. There's no SQL formatter tool that supports the Mako template manager. However, there is one that supports Jinja:
+(sqlfmt)[https://github.com/tconbeer/sqlfmt]. I didn't want to use Jinja just because there's a formatting tool that supports it. Hopefully, we will find a SQL formatting tool suitable for our use case, or we might consider contributing to this repository to write a Mako formatting connector.
```
Comment (mbabadi) [inline src/casp/cell_data_manager/README.md:117]: I -> We.

I suggest reading this README from top to bottom and making it a bit more formal. Thanks!

File: src/casp/cell_data_manager/README.md (hunk 5 above)
Comment (fedorgrab) [reply src/casp/cell_data_manager/README.md:33]: Lol, gotcha.:)

File: src/casp/cell_data_manager/sql/mako_helpers.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+    Results in::
+        select c.cell_type, i.dataset_id, c.cell_type
+        from `project-name-1.test_dataset.table_name`
+        ...
+    """
+    processed_columns = _process_column_names(column_names=column_names)
+    return f"select {', '.join(processed_columns)}" if processed_columns else "*"
+
+
+def where(filters: t.Optional[t.Dict[str, t.Any]]) -> str:
```
Comment (mbabadi) [inline src/casp/cell_data_manager/sql/mako_helpers.py:101]: How difficult would it be to implement `<...>__not_in` and `<...>__not_eq`?

**Review by mbabadi — CHANGES_REQUESTED** [review]
Excellent and high quality work Fedor!  I left a few comments and discussion points for you.

File: src/casp/cell_data_manager/README.md (hunk 7 above)
Comment (fedorgrab) [reply src/casp/cell_data_manager/README.md:117]: https://github.com/cellarium-ai/cellarium-cloud/issues/108 

I'll work on this pretty soon.

File: src/casp/cell_data_manager/sql/mako_helpers.py (hunk 8 above)
Comment (fedorgrab) [reply src/casp/cell_data_manager/sql/mako_helpers.py:101]: https://github.com/cellarium-ai/cellarium-cloud/issues/108

I'll work on this pretty soon.

**Review by mbabadi — APPROVED** [review]
Excellent, approved!

**Pull request merged**